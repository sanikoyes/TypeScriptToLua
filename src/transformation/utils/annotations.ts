import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum AnnotationKind {
    Extension = "extension",
    MetaExtension = "metaExtension",
    CustomConstructor = "customConstructor",
    CompileMembersOnly = "compileMembersOnly",
    NoResolution = "noResolution",
    PureAbstract = "pureAbstract",
    Phantom = "phantom",
    TupleReturn = "tupleReturn",
    LuaIterator = "luaIterator",
    LuaTable = "luaTable",
    NoSelf = "noSelf",
    NoSelfInFile = "noSelfInFile",
    Vararg = "vararg",
    ForRange = "forRange",
    Jsx = "jsx",
    JsxFrag = "jsxFrag",
}

export interface Annotation {
    kind: AnnotationKind;
    args: string[];
}

function createAnnotation(name: string, args: string[]): Annotation | undefined {
    const kind = Object.values(AnnotationKind).find(k => k.toLowerCase() === name.toLowerCase());
    if (kind !== undefined) {
        return { kind, args };
    }
}

export type AnnotationsMap = Map<AnnotationKind, Annotation>;

function collectAnnotations(source: ts.Symbol | ts.Signature, annotationsMap: AnnotationsMap): void {
    for (const tag of source.getJsDocTags()) {
        const annotation = createAnnotation(tag.name, tag.text?.map(p => p.text) ?? []);
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }
}

export function getSymbolAnnotations(symbol: ts.Symbol): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(symbol, annotationsMap);
    return annotationsMap;
}

export function getTypeAnnotations(type: ts.Type): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (type.symbol) collectAnnotations(type.symbol, annotationsMap);
    if (type.aliasSymbol) collectAnnotations(type.aliasSymbol, annotationsMap);

    return annotationsMap;
}

export function getNodeAnnotations(node: ts.Node): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    for (const tag of ts.getJSDocTags(node)) {
        const tagName = tag.tagName.text;
        const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }

    return annotationsMap;
}

export function getFileAnnotations(sourceFile: ts.SourceFile): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (sourceFile.statements.length > 0) {
        // Manually collect jsDoc because `getJSDocTags` includes tags only from closest comment
        const jsDoc = sourceFile.statements[0].jsDoc;
        if (jsDoc) {
            for (const tag of jsDoc.flatMap(x => x.tags ?? [])) {
                const tagName = tag.tagName.text;
                const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
                if (annotation) {
                    annotationsMap.set(annotation.kind, annotation);
                }
            }
        }
    }

    return annotationsMap;
}

export function getSignatureAnnotations(context: TransformationContext, signature: ts.Signature): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(signature, annotationsMap);

    // Function properties on interfaces have the JSDoc tags on the parent PropertySignature
    const declaration = signature.getDeclaration();
    if (declaration?.parent && ts.isPropertySignature(declaration.parent)) {
        const symbol = context.checker.getSymbolAtLocation(declaration.parent.name);
        if (symbol) {
            collectAnnotations(symbol, annotationsMap);
        }
    }

    return annotationsMap;
}

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getSignatureAnnotations(context, signature).has(AnnotationKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration?.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getTypeAnnotations(type).has(AnnotationKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.LuaIterator);
}

export function isVarargType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.Vararg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.ForRange);
}

export function getTagArgsFromComment(tag: ts.JSDocTag): string[] {
    if (tag.comment) {
        if (typeof tag.comment === "string") {
            return tag.comment.split(" ");
        } else {
            return tag.comment.map(part => part.text);
        }
    }

    return [];
}
