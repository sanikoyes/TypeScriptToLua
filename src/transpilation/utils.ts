import * as path from "path";
import * as resolve from "resolve";
import { SourceNode } from "source-map";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import * as lua from "../LuaAST";
import * as diagnosticFactories from "./diagnostics";

export interface EmitHost {
    directoryExists(path: string): boolean;
    fileExists(path: string): boolean;
    getCurrentDirectory(): string;
    readFile(path: string): string | undefined;
    writeFile: ts.WriteFileCallback;
}

interface BaseFile {
    code: string;
    sourceMap?: string;
    sourceFiles?: ts.SourceFile[];
}

export interface ProcessedFile extends BaseFile {
    fileName: string;
    luaAst?: lua.File;
    /** @internal */
    sourceMapNode?: SourceNode;
}

export interface EmitFile extends BaseFile {
    outputPath: string;
}

export const getConfigDirectory = (options: ts.CompilerOptions) =>
    options.configFilePath ? path.dirname(options.configFilePath) : process.cwd();

export function resolvePlugin(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        resolved = resolve.sync(query, { basedir, extensions: [".js", ".ts", ".tsx"] });
    } catch (err) {
        if (err.code !== "MODULE_NOT_FOUND") throw err;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            const tsNodePath = resolve.sync("ts-node", { basedir });
            const tsNode: typeof import("ts-node") = require(tsNodePath);
            tsNode.register({ transpileOnly: true });
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
