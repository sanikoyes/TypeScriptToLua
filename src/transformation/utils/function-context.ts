import * as ts from "typescript";
import { CompilerOptions } from "../../CompilerOptions";
import { TransformationContext } from "../context";
import { AnnotationKind, getFileAnnotations, getNodeAnnotations } from "./annotations";
import { findFirstNodeAbove, getAllCallSignatures, inferAssignedType } from "./typescript";

export enum ContextType {
    None,
    Void,
    NonVoid,
    Mixed,
}

function hasNoSelfAncestor(declaration: ts.Declaration): boolean {
    const scopeDeclaration = findFirstNodeAbove(
        declaration,
        (node): node is ts.SourceFile | ts.ModuleDeclaration => ts.isSourceFile(node) || ts.isModuleDeclaration(node)
    );

    if (!scopeDeclaration) {
        return false;
    } else if (ts.isSourceFile(scopeDeclaration)) {
        return getFileAnnotations(scopeDeclaration).has(AnnotationKind.NoSelfInFile);
    } else if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
        return true;
    } else {
        return hasNoSelfAncestor(scopeDeclaration);
    }
}

function getExplicitThisParameter(signatureDeclaration: ts.SignatureDeclaration): ts.ParameterDeclaration | undefined {
    return signatureDeclaration.parameters.find(
        param => ts.isIdentifier(param.name) && param.name.originalKeywordKind === ts.SyntaxKind.ThisKeyword
    );
}

export function getDeclarationContextType(
    { program }: TransformationContext,
    signatureDeclaration: ts.SignatureDeclaration
): ContextType {
    const thisParameter = getExplicitThisParameter(signatureDeclaration);
    if (thisParameter) {
        // Explicit 'this'
        return thisParameter.type && thisParameter.type.kind === ts.SyntaxKind.VoidKeyword
            ? ContextType.Void
            : ContextType.NonVoid;
    }

    // noSelf declaration on function signature
    if (getNodeAnnotations(signatureDeclaration).has(AnnotationKind.NoSelf)) {
        return ContextType.Void;
    }

    if (
        ts.isMethodSignature(signatureDeclaration) ||
        ts.isMethodDeclaration(signatureDeclaration) ||
        ts.isConstructSignatureDeclaration(signatureDeclaration) ||
        ts.isConstructorDeclaration(signatureDeclaration) ||
        (signatureDeclaration.parent && ts.isPropertyDeclaration(signatureDeclaration.parent)) ||
        (signatureDeclaration.parent && ts.isPropertySignature(signatureDeclaration.parent))
    ) {
        // Class/interface methods only respect @noSelf on their parent
        const scopeDeclaration = findFirstNodeAbove(
            signatureDeclaration,
            (n): n is ts.ClassLikeDeclaration | ts.InterfaceDeclaration =>
                ts.isClassDeclaration(n) || ts.isClassExpression(n) || ts.isInterfaceDeclaration(n)
        );

        if (scopeDeclaration === undefined) {
            return ContextType.NonVoid;
        }

        if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
            return ContextType.Void;
        }

        return ContextType.NonVoid;
    }

    // When using --noImplicitSelf and the signature is defined in a file targeted by the program apply the @noSelf rule.
    const options = program.getCompilerOptions() as CompilerOptions;
    if (options.noImplicitSelf && program.getRootFileNames().includes(signatureDeclaration.getSourceFile().fileName)) {
        return ContextType.Void;
    }

    // Walk up to find @noSelf or @noSelfInFile
    if (hasNoSelfAncestor(signatureDeclaration)) {
        return ContextType.Void;
    }

    return ContextType.NonVoid;
}

function reduceContextTypes(contexts: ContextType[]): ContextType {
    const reducer = (a: ContextType, b: ContextType) => {
        if (a === ContextType.None) {
            return b;
        } else if (b === ContextType.None) {
            return a;
        } else if (a !== b) {
            return ContextType.Mixed;
        } else {
            return a;
        }
    };

    return contexts.reduce(reducer, ContextType.None);
}

function getSignatureDeclarations(
    context: TransformationContext,
    signatures: readonly ts.Signature[]
): ts.SignatureDeclaration[] {
    return signatures.flatMap(signature => {
        const signatureDeclaration = signature.getDeclaration();
        let inferredType: ts.Type | undefined;
        if (
            ts.isMethodDeclaration(signatureDeclaration) &&
            ts.isObjectLiteralExpression(signatureDeclaration.parent) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            inferredType = context.checker.getContextualTypeForObjectLiteralElement(signatureDeclaration);
        } else if (
            (ts.isFunctionExpression(signatureDeclaration) || ts.isArrowFunction(signatureDeclaration)) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            // Infer type of function expressions/arrow functions
            inferredType = inferAssignedType(context, signatureDeclaration);
        }

        if (inferredType) {
            const inferredSignatures = getAllCallSignatures(inferredType);
            if (inferredSignatures.length > 0) {
                return inferredSignatures.map(s => s.getDeclaration());
            }
        }

        return signatureDeclaration;
    });
}

export function getFunctionContextType(context: TransformationContext, type: ts.Type): ContextType {
    if (type.isTypeParameter()) {
        type = type.getConstraint() ?? type;
    }

    if (type.isUnion()) {
        return reduceContextTypes(type.types.map(t => getFunctionContextType(context, t)));
    }

    const signatures = context.checker.getSignaturesOfType(type, ts.SignatureKind.Call);
    if (signatures.length === 0) {
        return ContextType.None;
    }

    const signatureDeclarations = getSignatureDeclarations(context, signatures);
    return reduceContextTypes(signatureDeclarations.map(s => getDeclarationContextType(context, s)));
}
