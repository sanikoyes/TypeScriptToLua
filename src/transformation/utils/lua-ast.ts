import * as ts from "typescript";
import { LuaTarget } from "../../CompilerOptions";
import * as lua from "../../LuaAST";
import { assert, castArray } from "../../utils";
import { TransformationContext } from "../context";
import { createExportedIdentifier, getIdentifierExportScope } from "./export";
import { peekScope, ScopeType, Scope } from "./scope";
import { transformLuaLibFunction } from "./lualib";
import { LuaLibFeature } from "../../LuaLib";

export type OneToManyVisitorResult<T extends lua.Node> = T | T[] | undefined;
export function unwrapVisitorResult<T extends lua.Node>(result: OneToManyVisitorResult<T>): T[] {
    return result === undefined ? [] : castArray(result);
}

export function createSelfIdentifier(tsOriginal?: ts.Node): lua.Identifier {
    return lua.createIdentifier("self", tsOriginal, undefined, "this");
}

export function createExportsIdentifier(): lua.Identifier {
    return lua.createIdentifier("____exports");
}

export function addToNumericExpression(expression: lua.Expression, change: number): lua.Expression {
    if (change === 0) return expression;

    const literalValue = getNumberLiteralValue(expression);
    if (literalValue !== undefined) {
        const newNode = lua.createNumericLiteral(literalValue + change);
        lua.setNodePosition(newNode, expression);
        return newNode;
    }

    if (lua.isBinaryExpression(expression)) {
        if (
            lua.isNumericLiteral(expression.right) &&
            ((expression.operator === lua.SyntaxKind.SubtractionOperator && expression.right.value === change) ||
                (expression.operator === lua.SyntaxKind.AdditionOperator && expression.right.value === -change))
        ) {
            return expression.left;
        }
    }

    return change > 0
        ? lua.createBinaryExpression(expression, lua.createNumericLiteral(change), lua.SyntaxKind.AdditionOperator)
        : lua.createBinaryExpression(expression, lua.createNumericLiteral(-change), lua.SyntaxKind.SubtractionOperator);
}

export function getNumberLiteralValue(expression?: lua.Expression) {
    if (!expression) return undefined;

    if (lua.isNumericLiteral(expression)) return expression.value;

    if (
        lua.isUnaryExpression(expression) &&
        expression.operator === lua.SyntaxKind.NegationOperator &&
        lua.isNumericLiteral(expression.operand)
    ) {
        return -expression.operand.value;
    }

    return undefined;
}

// Prefer use of transformToImmediatelyInvokedFunctionExpression to maintain correct scope. If you use this directly,
// ensure you push/pop a function scope appropriately to avoid incorrect vararg optimization.
export function createImmediatelyInvokedFunctionExpression(
    statements: lua.Statement[],
    result: lua.Expression | lua.Expression[],
    tsOriginal?: ts.Node
): lua.CallExpression {
    const body = [...statements, lua.createReturnStatement(castArray(result))];
    const flags = statements.length === 0 ? lua.FunctionExpressionFlags.Inline : lua.FunctionExpressionFlags.None;
    const iife = lua.createFunctionExpression(lua.createBlock(body), undefined, undefined, flags);
    return lua.createCallExpression(iife, [], tsOriginal);
}

export function createUnpackCall(
    context: TransformationContext,
    expression: lua.Expression,
    tsOriginal?: ts.Node
): lua.Expression {
    if (context.luaTarget === LuaTarget.Universal) {
        return transformLuaLibFunction(context, LuaLibFeature.Unpack, tsOriginal, expression);
    }

    const unpack =
        context.luaTarget === LuaTarget.Lua51 || context.luaTarget === LuaTarget.LuaJIT
            ? lua.createIdentifier("unpack")
            : lua.createTableIndexExpression(lua.createIdentifier("table"), lua.createStringLiteral("unpack"));

    return lua.createCallExpression(unpack, [expression], tsOriginal);
}

export function wrapInTable(...expressions: lua.Expression[]): lua.TableExpression {
    const fields = expressions.map(e => lua.createTableFieldExpression(e));
    return lua.createTableExpression(fields);
}

export function wrapInToStringForConcat(expression: lua.Expression): lua.Expression {
    if (
        lua.isStringLiteral(expression) ||
        lua.isNumericLiteral(expression) ||
        (lua.isBinaryExpression(expression) && expression.operator === lua.SyntaxKind.ConcatOperator)
    ) {
        return expression;
    }

    return lua.createCallExpression(lua.createIdentifier("tostring"), [expression]);
}

export function createHoistableVariableDeclarationStatement(
    context: TransformationContext,
    identifier: lua.Identifier,
    initializer?: lua.Expression,
    tsOriginal?: ts.Node
): lua.AssignmentStatement | lua.VariableDeclarationStatement {
    const declaration = lua.createVariableDeclarationStatement(identifier, initializer, tsOriginal);
    if (identifier.symbolId !== undefined) {
        const scope = peekScope(context);
        assert(scope.type !== ScopeType.Switch);

        if (!scope.variableDeclarations) {
            scope.variableDeclarations = [];
        }

        scope.variableDeclarations.push(declaration);
    }

    return declaration;
}

function hasMultipleReferences(scope: Scope, identifiers: lua.Identifier | lua.Identifier[]) {
    const scopeSymbols = scope.referencedSymbols;
    if (!scopeSymbols) {
        return false;
    }

    const referenceLists = castArray(identifiers).map(i => i.symbolId && scopeSymbols.get(i.symbolId));

    return referenceLists.some(symbolRefs => symbolRefs && symbolRefs.length > 1);
}

export function createLocalOrExportedOrGlobalDeclaration(
    context: TransformationContext,
    lhs: lua.Identifier | lua.Identifier[],
    rhs?: lua.Expression | lua.Expression[],
    tsOriginal?: ts.Node,
    overrideExportScope?: ts.SourceFile | ts.ModuleDeclaration
): lua.Statement[] {
    let declaration: lua.VariableDeclarationStatement | undefined;
    let assignment: lua.AssignmentStatement | undefined;

    const isFunctionDeclaration = tsOriginal !== undefined && ts.isFunctionDeclaration(tsOriginal);

    const identifiers = castArray(lhs);
    if (identifiers.length === 0) {
        return [];
    }

    const exportScope = overrideExportScope ?? getIdentifierExportScope(context, identifiers[0]);
    if (exportScope) {
        // exported
        if (!rhs) {
            return [];
        } else {
            assignment = lua.createAssignmentStatement(
                identifiers.map(identifier => createExportedIdentifier(context, identifier, exportScope)),
                rhs,
                tsOriginal
            );
        }
    } else {
        const scope = peekScope(context);
        const isTopLevelVariable = scope.type === ScopeType.File;

        if (context.isModule || !isTopLevelVariable) {
            if (!isFunctionDeclaration && hasMultipleReferences(scope, lhs)) {
                // Split declaration and assignment of identifiers that reference themselves in their declaration
                declaration = lua.createVariableDeclarationStatement(lhs, undefined, tsOriginal);
                if (rhs) {
                    assignment = lua.createAssignmentStatement(lhs, rhs, tsOriginal);
                }
            } else {
                declaration = lua.createVariableDeclarationStatement(lhs, rhs, tsOriginal);
            }

            if (!isFunctionDeclaration) {
                // Remember local variable declarations for hoisting later
                if (!scope.variableDeclarations) {
                    scope.variableDeclarations = [];
                }

                scope.variableDeclarations.push(declaration);
            }
        } else if (rhs) {
            // global
            assignment = lua.createAssignmentStatement(lhs, rhs, tsOriginal);
        } else {
            return [];
        }
    }

    if (isFunctionDeclaration) {
        // Remember function definitions for hoisting later
        const functionSymbolId = (lhs as lua.Identifier).symbolId;
        const scope = peekScope(context);
        if (functionSymbolId && scope.functionDefinitions) {
            const definitions = scope.functionDefinitions.get(functionSymbolId);
            if (definitions) {
                definitions.definition = declaration ?? assignment;
            }
        }
    }

    if (declaration && assignment) {
        return [declaration, assignment];
    } else if (declaration) {
        return [declaration];
    } else if (assignment) {
        return [assignment];
    } else {
        return [];
    }
}

export const createNaN = (tsOriginal?: ts.Node) =>
    lua.createBinaryExpression(
        lua.createNumericLiteral(0),
        lua.createNumericLiteral(0),
        lua.SyntaxKind.DivisionOperator,
        tsOriginal
    );
