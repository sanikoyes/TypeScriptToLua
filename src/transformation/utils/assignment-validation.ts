import * as ts from "typescript";
import { getOrUpdate } from "../../utils";
import { TransformationContext } from "../context";
import {
    unsupportedNoSelfFunctionConversion,
    unsupportedOverloadAssignment,
    unsupportedSelfFunctionConversion,
} from "./diagnostics";
import { ContextType, getFunctionContextType } from "./function-context";

// TODO: Clear if types are reused between compilations
const typeValidationCache = new WeakMap<ts.Type, Set<ts.Type>>();

export function validateAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    if (toType === fromType) {
        return;
    }

    if ((toType.flags & ts.TypeFlags.Any) !== 0) {
        // Assigning to un-typed variable
        return;
    }

    // Use cache to avoid repeating check for same types (protects against infinite loop in recursive types)
    const fromTypeCache = getOrUpdate(typeValidationCache, fromType, () => new Set());
    if (fromTypeCache.has(toType)) return;
    fromTypeCache.add(toType);

    validateFunctionAssignment(context, node, fromType, toType, toName);

    const fromTypeNode = context.checker.typeToTypeNode(fromType, undefined, undefined);
    const toTypeNode = context.checker.typeToTypeNode(toType, undefined, undefined);
    if (!fromTypeNode || !toTypeNode) {
        return;
    }

    if (
        (ts.isArrayTypeNode(toTypeNode) || ts.isTupleTypeNode(toTypeNode)) &&
        (ts.isArrayTypeNode(fromTypeNode) || ts.isTupleTypeNode(fromTypeNode))
    ) {
        // Recurse into arrays/tuples
        const fromTypeArguments = (fromType as ts.TypeReference).typeArguments;
        const toTypeArguments = (toType as ts.TypeReference).typeArguments;

        if (fromTypeArguments === undefined || toTypeArguments === undefined) {
            return;
        }

        const count = Math.min(fromTypeArguments.length, toTypeArguments.length);
        for (let i = 0; i < count; ++i) {
            validateAssignment(context, node, fromTypeArguments[i], toTypeArguments[i], toName);
        }
    }

    if (
        (toType.flags & ts.TypeFlags.Object) !== 0 &&
        (ts.isTypeLiteralNode(toTypeNode) ||
            ((toType as ts.ObjectType).objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0) &&
        toType.symbol &&
        toType.symbol.members &&
        fromType.symbol &&
        fromType.symbol.members
    ) {
        // Recurse into interfaces
        toType.symbol.members.forEach((toMember, escapedMemberName) => {
            if (fromType.symbol.members) {
                const fromMember = fromType.symbol.members.get(escapedMemberName);
                if (fromMember) {
                    const toMemberType = context.checker.getTypeOfSymbolAtLocation(toMember, node);
                    const fromMemberType = context.checker.getTypeOfSymbolAtLocation(fromMember, node);
                    const memberName = ts.unescapeLeadingUnderscores(escapedMemberName);
                    validateAssignment(
                        context,
                        node,
                        fromMemberType,
                        toMemberType,
                        toName ? `${toName}.${memberName}` : memberName
                    );
                }
            }
        });
    }
}

function validateFunctionAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    const fromContext = getFunctionContextType(context, fromType);
    const toContext = getFunctionContextType(context, toType);

    if (fromContext === ContextType.Mixed || toContext === ContextType.Mixed) {
        context.diagnostics.push(unsupportedOverloadAssignment(node, toName));
    } else if (fromContext !== toContext && fromContext !== ContextType.None && toContext !== ContextType.None) {
        context.diagnostics.push(
            toContext === ContextType.Void
                ? unsupportedNoSelfFunctionConversion(node, toName)
                : unsupportedSelfFunctionConversion(node, toName)
        );
    }
}
