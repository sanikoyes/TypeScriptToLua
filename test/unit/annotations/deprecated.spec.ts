import { annotationRemoved } from "../../../src/transformation/utils/diagnostics";
import * as util from "../../util";

test.each(["extension", "metaExtension"])("extension removed", extensionType => {
    util.testModule`
        declare class A {}
        /** @${extensionType} **/
        class B extends A {}
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("phantom removed", () => {
    util.testModule`
        /** @phantom **/
        namespace A {
            function nsMember() {}
        }
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("pureAbstract removed", () => {
    util.testModule`
        /** @pureAbstract */
        declare class ClassA {}
        class ClassB extends ClassA {}
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("forRange removed", () => {
    util.testModule`
        /** @forRange */
        declare function forRange(start: number, limit: number, step?: number): number[];
        for (const i of forRange(1, 10)) {}
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("vararg removed", () => {
    util.testModule`
        /** @vararg */
        type VarArg<T extends any[]> = T & { readonly __brand: unique symbol };
        function foo(...args: any[]) {}
        function vararg(...args: VarArg<any[]>) {
            foo(...args);
        }
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("luaiterator removed", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        /** @luaIterator */
        interface Iter extends Iterable<string> {}
        function luaIter(): Iter {
            let i = 0;
            return (() => arr[i++]) as any;
        }
        let result = "";
        for (let e of luaIter()) { result += e; }
        return result;
    `.expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

const tableLibClass = `
/** @luaTable */
declare class Table<K extends {} = {}, V = any> {
    length: number;
    constructor(notAllowed?: any);
    set(key?: K, value?: V, notAllowed?: any): void;
    get(key?: K, notAllowed?: any): V;
    other(): void;
}
declare let tbl: Table;
`;

test("LuaTable removed warning constructor", () => {
    util.testModule("const table = new Table();")
        .setTsHeader(tableLibClass)
        .expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("LuaTable removed warning property access length", () => {
    util.testFunction`
        return tbl.length;
    `
        .setTsHeader(tableLibClass)
        .expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("LuaTable removed warning property access get", () => {
    util.testFunction`
        return tbl.get("foo");
    `
        .setTsHeader(tableLibClass)
        .expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});

test("LuaTable deprecation warning property access set", () => {
    util.testFunction`
        return tbl.set("foo", 0);
    `
        .setTsHeader(tableLibClass)
        .expectDiagnosticsToMatchSnapshot([annotationRemoved.code]);
});
