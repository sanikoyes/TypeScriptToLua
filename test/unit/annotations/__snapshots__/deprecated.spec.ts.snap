// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuaTable deprecation warning property access set: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    return tbl:set(\\"foo\\", 0)
end
return ____exports"
`;

exports[`LuaTable deprecation warning property access set: diagnostics 1`] = `"main.ts(12,16): error TSTL: '@luaTable' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#luatable for more information."`;

exports[`LuaTable removed warning constructor: code 1`] = `
"require(\\"lualib_bundle\\");
____table = __TS__New(Table)"
`;

exports[`LuaTable removed warning constructor: diagnostics 1`] = `"main.ts(11,15): error TSTL: '@luaTable' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#luatable for more information."`;

exports[`LuaTable removed warning property access get: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    return tbl:get(\\"foo\\")
end
return ____exports"
`;

exports[`LuaTable removed warning property access get: diagnostics 1`] = `"main.ts(12,16): error TSTL: '@luaTable' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#luatable for more information."`;

exports[`LuaTable removed warning property access length: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    return tbl.length
end
return ____exports"
`;

exports[`LuaTable removed warning property access length: diagnostics 1`] = `"main.ts(12,16): error TSTL: '@luaTable' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#luatable for more information."`;

exports[`extension removed: code 1`] = `
"local lualib_bundle = require(\\"lualib_bundle\\");
local _ENV = setmetatable({}, { __index = lualib_bundle })
pcall(function() setfenv(1, _ENV) end)
B = __TS__Class()
B.name = \\"B\\"
__TS__ClassExtends(B, A)"
`;

exports[`extension removed: code 2`] = `
"local lualib_bundle = require(\\"lualib_bundle\\");
local _ENV = setmetatable({}, { __index = lualib_bundle })
pcall(function() setfenv(1, _ENV) end)
B = __TS__Class()
B.name = \\"B\\"
__TS__ClassExtends(B, A)"
`;

exports[`extension removed: diagnostics 1`] = `"main.ts(4,9): error TSTL: '@extension' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#extension for more information."`;

exports[`extension removed: diagnostics 2`] = `"main.ts(4,9): error TSTL: '@metaExtension' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#metaextension for more information."`;

exports[`forRange removed: code 1`] = `""`;

exports[`forRange removed: diagnostics 1`] = `"main.ts(4,25): error TSTL: '@forRange' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#forrange for more information."`;

exports[`luaiterator removed: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local arr = {\\"a\\", \\"b\\", \\"c\\"}
    local function luaIter(self)
        local i = 0
        return function() return arr[(function()
            local ____tmp = i
            i = ____tmp + 1
            return ____tmp
        end)() + 1] end
    end
    local result = \\"\\"
    return result
end
return ____exports"
`;

exports[`luaiterator removed: diagnostics 1`] = `"main.ts(10,23): error TSTL: '@luaIterator' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#luaiterator for more information."`;

exports[`phantom removed: code 1`] = `
"A = A or ({})
do
    local function nsMember(self)
    end
end"
`;

exports[`phantom removed: diagnostics 1`] = `"main.ts(3,9): error TSTL: '@phantom' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#phantom for more information."`;

exports[`pureAbstract removed: code 1`] = `
"local lualib_bundle = require(\\"lualib_bundle\\");
local _ENV = setmetatable({}, { __index = lualib_bundle })
pcall(function() setfenv(1, _ENV) end)
ClassB = __TS__Class()
ClassB.name = \\"ClassB\\"
__TS__ClassExtends(ClassB, ClassA)"
`;

exports[`pureAbstract removed: diagnostics 1`] = `"main.ts(4,22): error TSTL: '@pureAbstract' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#pureabstract for more information."`;

exports[`vararg removed: code 1`] = `
"function foo(self, ...)
end
function vararg(self, ...)
    foo(_G, ...)
end"
`;

exports[`vararg removed: diagnostics 1`] = `"main.ts(6,17): error TSTL: '@vararg' has been removed and will no longer have any effect.See https://typescripttolua.github.io/docs/advanced/compiler-annotations#vararg for more information."`;
